components/hybrid_engine/src/engine.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{KvEngine, MemoryEngine, Peekable, ReadOptions, Result, SyncMutable};
+
+use crate::snapshot::HybridSnapshot;
+
+/// This engine is structured with both a disk engine and an in-memory engine.
+/// The disk engine houses the complete database data, whereas the in-memory
+/// engine functions as a region cache, selectively caching certain regions to
+/// enhance read performance. For the regions that are cached, in-memory engine
+/// retains all data that has not been garbage collected.
+#[derive(Clone, Debug)]
+pub struct HybridEngine<EK, EM>
```

Comment:
I am thinking can we just add the region cache to rocksdb engine which impls `KvEngine` trait? 
It gains two benefits:
- you can switch on/off memory cache dynamically
- not make compile slower, avoid code bloat

Reply:
"Adding the region cache logic to rocksdb engine" also doesn't touch current code too much. It's just like use `HybridEngien` to replace `RocksEngine` and remove `RocksEngine`.

Reply:
I agree trait is more flexible. But I don't think there is another request in the near future.

Maybe compare the `make release` speed of before and after. If it doesn't introduce too much extra time, I'm okay with it. And still need to consider whether we need to switch it on/off dynamically.

Reply:
It means the implementation detail of memory engine will directly written into the relevant methods of RocksEngine.

Reply:
> I agree trait is more flexible. But I don't think there is another demand in the near future.
> 
> Maybe compare the `make release` speed of before and after. If it doesn't introduce too much extra time, I'm okay with it. And still need to consider whether we need to switch it on/off dynamically.

I will compare it later. I think dynamically switch on/off is easy to do in load/evict unit.

Reply:
Ok, it should be a swicher for those opening it at first place.

Reply:
> Maybe let's always use HybridEngine for raftstore, then it wouldn't have too much code bloats. When memory cache is disabled, just a single if branch to bypass doesn't affect performance.
> 
> And maybe rename `HybridEngine` to `RocksEngine`, and rename `RocksEngine` to something like `RocksEngineInner`. Then you don't need to change existing `xxx<RocksEngine>` to `xxx<HybridEngine>` all over the codespace

It's a good suggestion. We may discuss this with others at some time.

Reply:
> > I agree trait is more flexible. But I don't think there is another demand in the near future.
> > Maybe compare the `make release` speed of before and after. If it doesn't introduce too much extra time, I'm okay with it. And still need to consider whether we need to switch it on/off dynamically.
> 
> I will compare it later. I think dynamically switch on/off is easy to do in load/evict unit.

![image](https://github.com/tikv/tikv/assets/71589810/90a37c04-b540-4272-9410-cda38343b719)
Build period for this PR

![image](https://github.com/tikv/tikv/assets/71589810/6aea1fed-0818-47d1-9782-b9e82ee0378a)
Build period for master

At least no regression.